funcx_endpoint/funcx_endpoint/endpoint/endpoint_manager.py

```diff
@@ -0,0 +1,508 @@
+        log.info(f">>> Multi-Tenant Endpoint ID: {self._endpoint_uuid_str} <<<")
+
+        self._install_signal_handlers()
+
+        try:
+            self._event_loop()
+        except Exception:
+            log.exception("Unhandled exception; shutting down endpoint master")
+
+        ptitle = f"[shutdown in progress] {setproctitle.getproctitle()}"
+        setproctitle.setproctitle(ptitle)
+        self._command_stop_event.set()
+        self._kill_event.set()
+
+        os.killpg(os.getpgid(0), signal.SIGTERM)
+
+        proc_uid, proc_gid = os.getuid(), os.getgid()
+        for msg_prefix, signum in (
+            ("Signaling shutdown", signal.SIGTERM),
+            ("Forcibly killing", signal.SIGKILL),
+        ):
+            for pid, (uid, gid, uname, proc_args) in list(self._child_args.items()):
+                proc_ident = f"PID: {pid}, UID: {uid}, GID: {gid}, User: {uname}"
+                log.info(f"{msg_prefix} of user endpoint ({proc_ident}) [{proc_args}]")
+                try:
+                    os.setresgid(gid, gid, -1)
+                    os.setresuid(uid, uid, -1)
+                    os.killpg(os.getpgid(pid), signum)
+                except Exception as e:
+                    log.warning(
+                        f"User endpoint signal failed: {e} ({proc_ident}) [{proc_args}]"
+                    )
+                finally:
+                    os.setresuid(proc_uid, proc_uid, -1)
+                    os.setresgid(proc_gid, proc_gid, -1)
+
+            deadline = time.time() + 10
+            while self._child_args and time.time() < deadline:
+                time.sleep(0.5)
+                self.wait_for_children()
+
+        self._command.join(5)
+        log.info("Shutdown complete.\n---------- Endpoint Manager ends ----------\n\n")
+
+    def _event_loop(self):
+        self._command.start()
+
+        local_user_lookup = {}
+        try:
+            with open("local_user_lookup.json") as f:
+                local_user_lookup = json.load(f)
+        except Exception as e:
+            msg = (
+                f"Unable to load local users ({e.__class__.__name__}) {e}\n"
+                "  Will be unable to respond to any commands; update the lookup file"
+                f" and either restart (stop, start) or SIGHUP ({os.getpid()}) this"
+                " endpoint."
+            )
+            log.error(msg)
+
+        valid_method_name_re = re.compile(r"^[A-Za-z][0-9A-Za-z_]{0,99}$")
+        max_skew_s = 180  # 3 minutes; ignore commands with out-of-date timestamp
+        while not self._time_to_stop:
+            if self._wait_for_child:
+                self.wait_for_children()
+
+            try:
+                _command = self._command_queue.get(timeout=1.0)
+                d_tag, props, body = _command
+                if props.headers and props.headers.get("debug", False):
+                    body_log_b = re.sub(rb"://([^:]+):[^@]+@", rb"://\1:***@", body)
```

Comment:
This is very similar to the logic in __init__ above, in test_emitted, and the version in endpoint.py:start_endpoint.  That's a lot of non-DRY.  I think it's time to make a utility function to re-use the logic.  Since it's in the -service repo as well, probably better in -common repo but I hesitate to use that nowadays.

Reply:
Added to `utils/__init__.py`